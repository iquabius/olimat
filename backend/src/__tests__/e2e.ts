/**
 * Estes testes estão utilizando o mesmo serviço usado pra
 * desenvolvimento. Talvez seja melhor configurar um serviço
 * somente para testes. Atualmente se uma entidade for excluída
 * pela aplicação frontend, alguns testes podem falhar.
 * Acho que isso possa ser feito com um '.env.test'. Também
 * precisamos resetar esse serviço de alguma forma:
 * 'npx prisma seed --reset'
 */

// import our production apollo-server instance
import { server } from '../';
import { gql } from 'apollo-server-express';

import { startTestServer, toPromise } from './__utils';
import { City } from '../__generated__/prisma-client';

// O id muda toda vez que semeamos o serviço do prisma,
// então os testes com 'snapshot' falham.
const getCitiesQuery = gql`
  query getCities {
    cities {
      name
    }
  }
`;

const getCitiesWithIdsQuery = gql`
  query getCitiesWithIds {
    cities {
      id
      name
    }
  }
`;

const getCityQuery = gql`
  query getCity($id: ID!) {
    city(id: $id) {
      name
    }
  }
`;

describe('Backend - e2e', () => {
  let stop;
  let graphql;

  beforeEach(async () => {
    const testServer = await startTestServer(server);
    stop = testServer.stop;
    graphql = testServer.graphql;
  });

  afterEach(() => stop());

  test('gets list of cities', async () => {
    const res = await toPromise(
      graphql({
        query: getCitiesQuery,
      }),
    );

    expect(res).toMatchSnapshot();
  });

  test('gets a single city', async () => {
    // O id sempre muda, então primeiro buscamos uma lista de todas as cidades.
    // graphqlgen talvez gere um tipo que podemos usar aqui.
    const allCitiesRes: { data: { cities: City[] } } = await toPromise(
      graphql({
        query: getCitiesWithIdsQuery,
      }),
    );
    // Pegamos a primeira cidade.
    const firstCity: City = allCitiesRes.data.cities[0];

    // Buscamos a cidade usando o id da primeira cidade retornada.
    // Isso parece redundante, parece que estamos testando o Prisma,
    // que não deveria ser nossa problema.
    const res = await toPromise(graphql({ query: getCityQuery, variables: { id: firstCity.id } }));

    expect(res).toMatchSnapshot();
  });
});
