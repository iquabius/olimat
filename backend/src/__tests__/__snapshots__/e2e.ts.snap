// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend - e2e gets a single city 1`] = `
Object {
  "data": Object {
    "city": Object {
      "name": "Sinop",
    },
  },
}
`;

exports[`Backend - e2e gets list of cities 1`] = `
Object {
  "data": Object {
    "cities": Array [
      Object {
        "name": "Sinop",
      },
      Object {
        "name": "CÃ¡ceres",
      },
      Object {
        "name": "Barra do Bugres",
      },
    ],
  },
}
`;
